ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs

```diff
@@ -134,3 +148,114 @@ prop_uniformSchedule = do
+-- from advancing, interrupting the test at this point should cause the
+-- immutable tip to be too far behind the last genesis window of the honest
+-- chain.
+
+-- | Test that the leashing attacks do not delay the immutable tip
+--
+-- This test is expected to fail because we don't test a genesis implementation
+-- yet.
+prop_leashingAttackStalling :: QC.Gen QC.Property
+prop_leashingAttackStalling = QC.expectFailure <$> do
+  genesisTest <- genGenesisTest
+  schedule <- genLeashingSchedule genesisTest
+  pure $
+    runSimOrThrow $
+    runTest schedulerConfig genesisTest schedule $
+    exceptionCounterexample $
+    makeProperty genesisTest (length (peerIds schedule) - 1)
+
+  where
+    schedulerConfig = (noTimeoutsSchedulerConfig scheduleConfig)
+      { scEnableGdd = True
+      , scTrace = False
+      }
+
+    scheduleConfig = defaultPointScheduleConfig
+
+    -- | Produces schedules that might cause the node under test to stall.
+    --
+    -- This is achieved by dropping random points from the schedule of each peer
+    genLeashingSchedule :: GenesisTest -> QC.Gen PointSchedule
+    genLeashingSchedule genesisTest =
+      unsafeMapSuchThatJust $ do
+        (honest, advs0) <- genUniformSchedulePoints genesisTest
+        advs <- mapM dropRandomPoints advs0
+        pure $ fromSchedulePoints (mkPeers honest advs)
+
+    dropRandomPoints :: [(DiffTime, SchedulePoint)] -> QC.Gen [(DiffTime, SchedulePoint)]
+    dropRandomPoints ps = do
+      let lenps = length ps
+      dropCount <- QC.choose (0, max 1 $ div lenps 5)
+      let dedup = map head . group
+      is <- fmap (dedup . sort) $ replicateM dropCount $ QC.choose (0, lenps - 1)
+      pure $ dropElemsAt ps is
+
+    dropElemsAt :: [a] -> [Int] -> [a]
+    dropElemsAt xs [] = xs
+    dropElemsAt xs (i:is) =
+      let (ys, zs) = splitAt i xs
+       in ys ++ dropElemsAt (drop 1 zs) is
+
+-- | Test that the leashing attacks do not delay the immutable tip after. The
+-- immutable tip needs to be advanced enough when the honest peer has offered
+-- all of its ticks.
+--
+-- This test is expected to fail because we don't test a genesis implementation
+-- yet.
+--
+-- See Note [Leashing attacks]
+prop_leashingAttackTimeLimited :: QC.Gen QC.Property
+prop_leashingAttackTimeLimited = QC.expectFailure <$> do
+  genesisTest <- genGenesisTest
+  schedule <- genTimeLimitedSchedule genesisTest
+  pure $
+    runSimOrThrow $
+    runTest schedulerConfig genesisTest schedule $
+    exceptionCounterexample $
+    makeProperty genesisTest (length (peerIds schedule) - 1)
+
+  where
+    schedulerConfig = (noTimeoutsSchedulerConfig scheduleConfig)
+      { scEnableGdd = True
+      , scTrace = False
+      }
+
+    scheduleConfig = defaultPointScheduleConfig
+
+    -- | A schedule which doesn't run past the last event of the honest peer
+    genTimeLimitedSchedule :: GenesisTest -> QC.Gen PointSchedule
+    genTimeLimitedSchedule genesisTest =
+      unsafeMapSuchThatJust $ do
```

Comment:
this has the effect that the leader schedule will not be regenerated when the point schedule produces `Nothing`. unsure whether that's relevant for this case or undesirable

Reply:
When does the point schedule produce Nothing?

Reply:
FTR we decided to replace `Maybe` with `error` since it will never happen.